File: SEM5/FLCD/lab2/lab2_ST_cpp/lab2_ST_cpp/SymbolTable.cpp
#include "SymbolTable.h"

// Constructor
SymbolTable::SymbolTable(int size) : hashTable(size) {}

// Find by position
std::string SymbolTable::findByPos(Pair pos) {
    return hashTable.findByPosition(pos);
}

// Get hash table
HashTable SymbolTable::getHashTable() {
    return hashTable;
}

// Get size
int SymbolTable::getSize() {
    return hashTable.getSize();
}

// Find position of term
Pair SymbolTable::findPositionOfTerm(const std::string& term) {
    return hashTable.findPositionOfTerm(term);
}

// Check if the term exists
bool SymbolTable::containsTerm(const std::string& term) {
    return hashTable.containsTerm(term);
}

// Add a term
bool SymbolTable::add(const std::string& term) {
    return hashTable.add(term);
}

File: SEM5/FLCD/lab2/lab2_impl_docu_Goian_Sergiu/lab2_ST_Hashtable/src/SymbolTable.java
import java.util.ArrayList;

public class SymbolTable {
    private HashTable identifierTable;
    private HashTable constantTable;

    public SymbolTable(int initialSize) {
        this.identifierTable = new HashTable(initialSize);
        this.constantTable = new HashTable(initialSize);
    }

    public String findByPosition(Pair position, boolean isConstant) {
        try {
            return isConstant ? constantTable.findByPos(position) : identifierTable.findByPos(position);
        } catch (IndexOutOfBoundsException e) {
            System.err.println("Invalid position: " + position);
            return null;
        }
    }

    public HashTable getIdentifierTable() {
        return identifierTable;
    }

    public HashTable getConstantTable() {
        return constantTable;
    }

    public int getIdentifierSize() {
        return identifierTable.getSize();
    }

    public int getConstantSize() {
        return constantTable.getSize();
    }

    public Pair findPositionOfTerm(String term, boolean isConstant) {
        return isConstant ? constantTable.findPositionOfTerm(term) : identifierTable.findPositionOfTerm(term);
    }

    public boolean containsTerm(String term, boolean isConstant) {
        return isConstant ? constantTable.containsTerm(term) : identifierTable.containsTerm(term);
    }

    public boolean addTerm(String term, boolean isConstant) {
        return isConstant ? constantTable.add(term) : identifierTable.add(term);
    }

    public boolean removeTerm(String term, boolean isConstant) {
        return isConstant ? constantTable.remove(term) : identifierTable.remove(term);
    }

    public void addIdentifier(String identifier) {
        if (!addTerm(identifier, false)) {
            System.out.println("Identifier already exists: " + identifier);
        }
    }

    public void addConstant(String constant) {
        if (!addTerm(constant, true)) {
            System.out.println("Constant already exists: " + constant);
        }
    }
}
